doc/specification/language-proposal.md

```diff
@@ -0,0 +1,1518 @@
+
+When interpreted or compiled into an Arbor mechanism, _R_ and operations upon it will necessarily be approximated in some manner; IEEE 754 semantics are a reasonable expectation, with NaN representing ⊥, but the interpreter or transpiler is permitted to perform algebraic simplifications including:
+
+1. Reordering of associative operations.
+2. Constant folding.
+3. Algebraic elimination of terms (which may change the value of some expressions from ⊥ to a value in ℝ ∪ {-∞, ∞}, or from an infinite value to a finite value).
+
+
+### Records
+<a id="records"/>
+
+A record is a labelled unordered tuple of values which are either quantities or records themselves. A record may not have two fields of the same name. A record type specification has the syntax:
+
+> _record-type-expr_ ::= `{` _field-spec_* `}`
+>
+> _field-spac_ ::= ***symbol*** _type-assertion_ `;`
+>
+> _type-assertion_ ::= `:` _type-expr_
+
+A record type _R_ with fields _fᵢ_ of type _Tᵢ_ is a supertype of a record type _S_ with fields _gⱼ_ of type _Uⱼ_ if for every _i_ there exists a _j_ such that _fᵢ_ and _gⱼ_ have the same identifier and _Tᵢ_ is the same type as or a supertype of _Uⱼ_. If _R_ is a supertype of _S_ and _S_ is a supertype of _R_, then _R_ and _S_ are the same type. Note that the order of fields in a _record-type_ is arbitrary.
+
+Correspondingly, record literals have the syntax:
+
+> _record-literal_ ::= `{` _field-defn_* `}`
+>
+> _field-defn_ ::= ***symbol*** _type-assertion_? `=` _expression_ `;` | _reaction_ `;`
+
+For field definitions, if there is no type assertion given, the type of the field is deduced from the expression on the right hand side.
+
+The field names constituting the left hand side of the field definitions have the _record scope_ of the record value. The defining expressions on the right hand side remain in an expression context, so that identifiers with the same symbol as a record field will _not_ refer to that record. For example, in the expression
+
+```
+let a = 3 m;
+let r = { a = 4; b = a; };
+r.b
+```
+
+`r` has type `{ a: real; b: length; }` and is bound to the value `{ a = 4; b = 3 m; }`, because the `a` in the right hand side of `b = a` is in expression context, and so is bound to to the value 3 m as given in the outer `let`.
+
+Field values in a record are accessed via a qualified identifier (see [Qualified identifiers](#qualified-identifiers) above), or by immediate field access (see [Record expressions](#record-expressions) below), or brought into scope via a `with` binding (see [Value bindings](#value-bindings), below).
+
+```
+let r = { a = 4; }; r.a   # qualified identifier evaluates to value 4
+let x = { a = 4; }.a; x   # immediate field access evaluates to value 4
+with { a = 4; }; a        # a is locally bound to the value 4
```

Comment:
Especially for `with` bindings -- which potentially introduce lots of bound symbols -- I would consider an explicitly delimited scope helpful.

Reply:
One can always use parentheses, if one wants to!
I want to avoid having matched delimiters around the final expression part, so that e.g. five `let`s don't give us five tailing close-scope symbols.